content/docs/installation/features.md

```diff
@@ -0,0 +1,62 @@
+---
+title: Features
+description: cert-manager feature policy
+---
+
+New cert-manager features and functionality is often initially implemented behind a feature gate. This is so as to not break users with functionality that has not yet been tested in production as well as to give us a chance to remove or modify API fields and functionality following user feedback.
+
+We have alpha and beta features. We do not aim to keep any of the features in alpha or beta stage indefinitely. Feature gating should only be used for functionality that can eventually be turned on by default for all users (GA).
```

Comment:
question: I think the way in which some features currently behind gates are implemented might break this?

e.g. ValidateCAA is feature gated, but it's [also](https://github.com/cert-manager/cert-manager/blob/b0caedffd915eab1a52e06473dc2b5b527b84460/pkg/controller/acmechallenges/sync.go#L149) a toggle on whether or not to do the validation. We might want to graduate ValidateCAA to GA (or not, it's just an example) but it would seem reasonable for users to want to disable it.

In fact, for this specific feature I'd imagine users would want to opt-in rather than opt-out (and that might even be required so that we don't break people).

Considering this:

> Feature gating should only be used for functionality that can eventually be turned on by default for all users (GA).

For this to be true, we'd have to add a new flag "--validate-caa" which is required _as well as the feature gate_, right? (Or some other way of toggling the feature on in addition to the feature gate)

Reply:
> For this to be true, we'd have to add a new flag "--validate-caa" which is required as well as the feature gate, right? (Or some other way of toggling the feature on in addition to the feature gate)

Yes, I think that the mechanism for marking a piece of functionality as alpha/beta should be separate from ways how something that should always be opt-in gets toggled on/off. I think that's better than mixing the two up because it simplifies the feature flag mechanism that is already complicated, is how it's implemented in kube and would appear more consistent to users once it's in GA.

It has the downside that when some functionality X that should always be opt-in is in alpha, using it might look like
```
go run <cert-manager-controller> \
 --features-gates=<optional-functionality-x>=true \
 --<enable-optional-functionality-x>
```
but I think this inconvenience is acceptable so that in GA we can have

```
go run <cert-manager-controller> 
 --<enable-optional-functionality-a>
 --<enable-optional-functionality-b>
 --<enable-optional-functionality-x>
```
and not
```
go run <cert-manager-controller> 
 --<enable-optional-functionality-a>
 --<enable-optional-functionality-b>
 --feature-gates=<enable-optional-functionality-x> // X is GA same as A and B, but gets toggled on via --feature-gates flag because it once was an alpha feature
```

I think it's ok if enabling an alpha feature is a bit clunky, perhaps it'd even encourage folks to nudge us to move it through to graduation.

Let me know what you think.

Reply:
> But since the features are feature gated

😆 

yeah, I think in some cases it might not be possible to know whether something will need to opt-in. For example with Gateway, we originally assumed that those CRDs will at some point become part of core Kubernetes.